hkmc2/shared/src/test/mlscript/nofib/NofibPrelude.mls

```diff
@@ -1,12 +1,15 @@
 
-fun nofibPrint(x) = globalThis.process.stdout.write(globalThis.JSON.stringify(x) + "\n")
-fun nofibError(e) = globalThis.eval("throw new Error('" + e + "')")
-fun nofibBigInt(i) = globalThis.BigInt(i)
+declare type Array
+set Array.prototype.toString = globalThis.eval("(function() { return '(' + Array.prototype.join.call(this, ',') + ')'; })")
```

Comment:
WOW. Is this really necessary?

Reply:
Yeah I used this gross hack to override `toString` for Arrays because otherwiss, some outputs of list of tuples become weird... for example in [this diff for `mandel.mls`](https://github.com/hkust-taco/mlscript/pull/251/commits/198c05da1f053ad8d6fecef96ab2ec59a416cb72#diff-06e58e019cdd717b37cc60596be892ea083b144b7d4fd21870d09eb129612e73L74), the list of tuples (`[(1,2), (2,3)]`) will be displayed as `[1,2,2,3]` without this..

Reply:
Wow that's _bad_. I can't believe people in JS and TS live with arrays/tuples that do not have a useful `toString` implementation when nested...

Reply:
Yes I just checked and confirmed that the default JS array `toString()` method gives outputs [that look like flattened...](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toString#:~:text=Array.prototype.toString%20recursively%20converts%20each%20element%2C%20including%20other%20arrays%2C%20to%20strings.%20Because%20the%20string%20returned%20by%20Array.prototype.toString%20does%20not%20have%20delimiters%2C%20nested%20arrays%20look%20like%20they%20are%20flattened.)

```javascript
> [[2,3], [1]].toString()
'2,3,1'
```

Maybe I can also use `util.inspect`, which gives more reasonable output shown below. But I think in `NofibPrelude.mls`, if I want to override `Array.prototype.toString`, I still have to use `eval` because `this` is not available to arrow functions generated from mls..
```javascript
> util.inspect([[2,3], [1]])
'[ [ 2, 3 ], [ 1 ] ]'
```

Reply:
Yeah we don't have the crazy `this` semantics of JS. We'll need to add a syntax that explicitly binds `this` when needed, in cases like this.

For now it's fine to use your hack, but this is of course not a real solution, as other parts of the program and libraries written in JS might rely on this behavior.

We could use a wrapper class that properly implements `toString`. `[1, 2, 3]` would be compiled to `new Tuple(1, 2, 3)`:

```js
> class Tuple extends Array { constructor(...xs) { super(...xs) } }
undefined
> new Tuple(1, 2, 3)
Tuple(3) [ 1, 2, 3 ]
```

But this could come in a separate PR.

Reply:
Thanks I see. So it seems that to enable this we need to have support for `super` and for `extends` in the js backend.